subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -385,256 +357,1118 @@ static int hfp_ag_send_data(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
+			ag_ind[BT_HFP_AG_SERVICE_IND].connector, ag_ind[BT_HFP_AG_SERVICE_IND].max,
+			ag_ind[BT_HFP_AG_CALL_IND].name, ag_ind[BT_HFP_AG_CALL_IND].min,
+			ag_ind[BT_HFP_AG_CALL_IND].connector, ag_ind[BT_HFP_AG_CALL_IND].max,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].name, ag_ind[BT_HFP_AG_CALL_SETUP_IND].min,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].connector,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].max, ag_ind[BT_HFP_AG_CALL_HELD_IND].name,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].min,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].connector,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].max, ag_ind[BT_HFP_AG_SIGNAL_IND].name,
+			ag_ind[BT_HFP_AG_SIGNAL_IND].min, ag_ind[BT_HFP_AG_SIGNAL_IND].connector,
+			ag_ind[BT_HFP_AG_SIGNAL_IND].max, ag_ind[BT_HFP_AG_ROAM_IND].name,
+			ag_ind[BT_HFP_AG_ROAM_IND].min, ag_ind[BT_HFP_AG_ROAM_IND].connector,
+			ag_ind[BT_HFP_AG_ROAM_IND].max, ag_ind[BT_HFP_AG_BATTERY_IND].name,
+			ag_ind[BT_HFP_AG_BATTERY_IND].min, ag_ind[BT_HFP_AG_BATTERY_IND].connector,
+			ag_ind[BT_HFP_AG_BATTERY_IND].max);
+	} else {
+		err = hfp_ag_send_data(ag, NULL, NULL, "\r\n+CIND:%d,%d,%d,%d,%d,%d,%d\r\n",
+					 ag->indicator_value[BT_HFP_AG_SERVICE_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_SETUP_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_HELD_IND],
+					 ag->indicator_value[BT_HFP_AG_SIGNAL_IND],
+					 ag->indicator_value[BT_HFP_AG_ROAM_IND],
+					 ag->indicator_value[BT_HFP_AG_BATTERY_IND]);
+	}
+
+	return err;
+}
+
+static void bt_hfp_ag_set_in_band_ring(struct bt_hfp_ag *ag, void *user_data)
+{
+	bool is_inband_ringtone;
+
+	hfp_ag_lock(ag);
+	is_inband_ringtone = (ag->ag_features & BT_HFP_AG_FEATURE_INBAND_RINGTONE) ? true : false;
+	hfp_ag_unlock(ag);
+
+	if (is_inband_ringtone && !atomic_test_bit(ag->flags, BT_HFP_AG_INBAND_RING)) {
+		int err = hfp_ag_send_data(ag, NULL, NULL, "\r\n+BSIR:1\r\n");
+
+		atomic_set_bit_to(ag->flags, BT_HFP_AG_INBAND_RING, err == 0);
+	}
+}
+
+static void bt_hfp_ag_slc_connected(struct bt_hfp_ag *ag, void *user_data)
+{
+	bt_hfp_ag_set_state(ag, BT_HFP_CONNECTED);
+	(void)hfp_ag_next_step(ag, bt_hfp_ag_set_in_band_ring, NULL);
+}
+
+static int bt_hfp_ag_cmer_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+{
+	uint32_t number;
+	int err;
+	static const uint32_t command_line_prefix[] = {3, 0, 0};
+
+	if (!is_char(buf, '=')) {
+		return -ENOTSUP;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(command_line_prefix); i++) {
+		err = get_number(buf, &number);
+		if (err != 0) {
+			return -ENOTSUP;
+		}
+
+		if (!is_char(buf, ',')) {
+			return -ENOTSUP;
+		}
+
+		if (command_line_prefix[i] != number) {
+			return -ENOTSUP;
+		}
+	}
+
+	err = get_number(buf, &number);
+	if (err != 0) {
+		return -ENOTSUP;
+	}
+
+	if (!is_char(buf, '\r')) {
+		return -ENOTSUP;
+	}
+
+	if (number == 1) {
+		atomic_set_bit(ag->flags, BT_HFP_AG_CMER_ENABLE);
+		if ((ag->hf_features & BT_HFP_HF_FEATURE_3WAY_CALL) &&
+		    (ag->ag_features & BT_HFP_AG_FEATURE_3WAY_CALL)) {
```

Comment:
And here there's no locking at all. It's really not clear to me when it's needed or not.